### ISSUE TITLE
TC segmentation fault before validation

### ISSUE THREAD

**Issue by maawad** [gunrock/gunrock#483]
**Summary:**
TC crashes before validation.

**Reproducibility:**
Always.

**To Reproduce:**
1. dev-refactor branch
2. run make in `gunrock/tests/tc`
3. run `./bin/tc_main_10.0_x86_64  market /data/gunrock_dataset/large/coAuthorsDBLP/coAuthorsDBLP.mtx`

**Actual Results:**
Crashes just before validating the results (probably [here](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/tc/tc_app.cu#L112)).
```
mawad@luigi:~/gunrock/tests/tc$ ./bin/tc_main_10.0_x86_64  market /data/gunrock_dataset/large/coAuthorsDBLP/coAuthorsDBLP.mtxLoading Matrix-market coordinate-formatted graph ...
  Reading meta data from /data/gunrock_dataset/large/coAuthorsDBLP/coAuthorsDBLP.mtx.meta
  Reading edge lists from /data/gunrock_dataset/large/coAuthorsDBLP/coAuthorsDBLP.mtx.coo_edge_pairs
  Assigning 1 to all 977676 edges
  Substracting 1 from node Ids...
  Edge doubleing: 977676 -> 1955352 edges
  graph loaded as COO in 0.069036s.
Converting 299067 vertices, 1955352 directed edges ( ordered tuples) to CSR format...Done (0s).
Degree Histogram (299067 vertices, 1955352 edges):
    Degree 0: 0 (0.000000 %)
    Degree 2^0: 41376 (13.835027 %)
    Degree 2^1: 102633 (34.317728 %)
    Degree 2^2: 87368 (29.213521 %)
    Degree 2^3: 42746 (14.293118 %)
    Degree 2^4: 17441 (5.831804 %)
    Degree 2^5: 6019 (2.012592 %)
    Degree 2^6: 1324 (0.442710 %)
    Degree 2^7: 156 (0.052162 %)
    Degree 2^8: 4 (0.001337 %)

__________________________
CPU_Reference: start
CPU_Reference: done
__________________________
Run CPU Reference Avg. in 1 iterations elapsed: 133.892059 ms
==============================================
 advance-mode=LB
Using advance mode LB
Using filter mode CULL
__________________________
--------------------------
Run 0 elapsed: 3.429890 ms, #iterations = 1
Segmentation fault (core dumped)
```
**System Configuration:**
Luigi.
GPU: K40c
CPU: Intel(R) Xeon(R) CPU E5-2637 v2 @ 3.50GHz
OS: Ubuntu 16.04.5 LTS
NVIDIA-SMI 410.72
Driver Version: 410.72
CUDA Version: 10.0
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.10) 5.4.0 20160609

**Comment by maawad** [comment]
Okay, I think I see where the problem is here.
The intersection kernel is counting intersections per [**edge**](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/oprtr/intersection/kernel.cuh#L237), but the reference code is counting per [**vertex**](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/tc/tc_test.cuh#L150). Also the [validation code](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/tc/tc_test.cuh#L213) is commented out.

 Changing this [allocation](https://github.com/gunrock/gunrock/blob/dev-refactor/gunrock/app/tc/tc_app.cu#L74) to be `VertexT *h_tc_counts = new VertexT[graph.edges];` fixes the problem (same output size as GPU code). Obviously the reference code needs to be changed for validation. 
@sgpyc @neoblizz @porumbes

**Comment by jowens** [comment]
@Laurawly please help with this

**Comment by Laurawly** [comment]
@jowens Lemme take a look

**Comment by Laurawly** [comment]
@jowens @maawad @neoblizz Solved by PR #490 
@maawad Note that if you're running TC on mario, BaseMakefile.mk needs to be changed to compile with SM35/37.
```laurawly@mario:~/workplace/gunrock/tests/tc$ ./run.sh
Loading Matrix-market coordinate-formatted graph ...
  Reading meta data from ../../dataset/small/data_sm.mtx.meta
  Reading edge lists from ../../dataset/small/data_sm.mtx.coo_edge_pairs
  Assigning 1 to all 10 edges
  Substracting 1 from node Ids...
  Edge doubleing: 10 -> 20 edges
  graph loaded as COO in 0.029310s.
Converting 5 vertices, 20 directed edges ( ordered tuples) to CSR format...Done (0s).
Degree Histogram (5 vertices, 20 edges):
    Degree 0: 0 (0.000000 %)
    Degree 2^0: 0 (0.000000 %)
    Degree 2^1: 0 (0.000000 %)
    Degree 2^2: 5 (100.000000 %)

__________________________
CPU_Reference: start
CPU_Reference: done
__________________________
Run CPU Reference Avg. in 10 iterations elapsed: 0.000310 ms
==============================================
 advance-mode=LB
Using advance mode LB
Using filter mode CULL
__________________________
--------------------------
Run 0 elapsed: 0.066996 ms, #iterations = 1
__________________________
--------------------------
Run 1 elapsed: 0.037193 ms, #iterations = 1
__________________________
--------------------------
Run 2 elapsed: 0.037909 ms, #iterations = 1
__________________________
--------------------------
Run 3 elapsed: 0.037909 ms, #iterations = 1
__________________________
--------------------------
Run 4 elapsed: 0.036001 ms, #iterations = 1
__________________________
--------------------------
Run 5 elapsed: 0.038147 ms, #iterations = 1
__________________________
--------------------------
Run 6 elapsed: 0.037909 ms, #iterations = 1
__________________________
--------------------------
Run 7 elapsed: 0.041008 ms, #iterations = 1
__________________________
--------------------------
Run 8 elapsed: 0.040054 ms, #iterations = 1
__________________________
--------------------------
Run 9 elapsed: 0.036001 ms, #iterations = 1
Validate_Results
Triangle Counting Validity: PASS
PASS
[TC] finished.
 avg. elapsed: 0.040913 ms
 iterations: 1
 min. elapsed: 0.036001 ms
 max. elapsed: 0.066996 ms
 load time: 48.768 ms
 preprocess time: 4320.770000 ms
 postprocess time: 0.354052 ms
 total time: 4322.397947 ms
